Support for 2D stencils (issue #781)

andrewcoughtrie opened the pull request.

andrewcoughtrie commented:
There is currently no code for the stub generator for the CROSS2D stencil, this will be done in a separate issue to avoid making the changes any more extensive.

andrewcoughtrie commented:
I can't seem to get the codecov patch check to be successful.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -1519,28 +1527,52 @@ def _declare_maps_invoke(self, parent):
+                parent.add(DeclGen(parent, datatype="integer",
+                                   kind=api_config.default_kind["integer"],
+                                   pointer=True,
+                                   entity_decls=[self.dofmap_name(symtab, arg) +
```
This line is >79 chars

andrewcoughtrie replied:
Fixed

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -1519,28 +1527,52 @@ def _declare_maps_invoke(self, parent):
-
             stencil_type = arg.descriptor.stencil['type']
+            if stencil_type == "cross2d":
+                parent.add(UseGen(parent, name="stencil_2D_dofmap_mod", only=True,
```
This line is > 79 chars

andrewcoughtrie replied:
Fixed

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -1519,28 +1527,52 @@ def _declare_maps_invoke(self, parent):
+                                                                       arg) +
+                                                 "(:) => null()"]))
+
             if stencil_type == "xory1d":
```
Can you merge this if-block with the previous one?

andrewcoughtrie replied:
I've merged it to an extent, it still requires two if blocks but one is nested inside the other now and only required for stencils types other than cross2d.

arporter commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -1099,10 +1104,10 @@ and the following stencil (with ``extent=2``):
   |   |   | 8 |   |   |
+  | 3 | 2 | 1 | 6 | 7 |
+  |   |   | 4 |   |   |
+  |   |   | 5 |   |   |
```
If I've understood the changes correctly, it seems that the way one of the existing stencil types is specified has changed too? Please could you update the rules on Subroutine arguments section (p.66 of the User Guide).

arporter replied:
In particular, I see that the stencil-size argument can now be an array lookup.

andrewcoughtrie replied:
I've updated the info on page 66 to give more detail on the kernel subroutine arguments and their purpose.

arporter commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -423,6 +432,21 @@ def stencil_unknown_extent(self, arg, var_accesses=None):
 
+    @abc.abstractmethod
+    def stencil_2d_unknown_extent(self, arg, var_accesses=None):
+        '''Add stencil information to the argument list associated with the
```
Please extend this docstring to make it clear how this differs from `stencil_unknown_extent`.

andrewcoughtrie replied:
Added identifier to distinguish from stencil_unknown_extent

arporter commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -455,6 +479,22 @@ def stencil(self, arg, var_accesses=None):
 
+    @abc.abstractmethod
+    def stencil_2d(self, arg, var_accesses=None):
+        '''Add general stencil information associated with the argument 'arg'
```
This docstring is currently identical to that for `stencil()`. Presumably it shouldn't be?

andrewcoughtrie replied:
Changed to reflect 2D stencil info

arporter commented on src/psyclone/domain/lfric/kern_call_acc_arg_list.py:
```diff
@@ -113,6 +113,60 @@ def stencil(self, arg, var_accesses=None):
+            :py:class:`psyclone.core.access_info.VariablesAccessInfo`
+
+        '''
+        from psyclone.dynamo0p3 import DynStencils
```
Since this implementation is identical (I think?) to that for `stencil()`, I'd just call `stencil()` here with a note to that effect.

andrewcoughtrie replied:
Done.

arporter commented on src/psyclone/domain/lfric/kern_call_acc_arg_list.py:
```diff
@@ -113,6 +113,60 @@ def stencil(self, arg, var_accesses=None):
+
+        '''
+        # The extent is not specified in the metadata so pass the value in
+        from psyclone.dynamo0p3 import DynStencils
```
Similarly, this can just call `stencil_unknown_extent()` I think?

andrewcoughtrie replied:
Done.

arporter commented on src/psyclone/domain/lfric/kern_call_arg_list.py:
```diff
@@ -237,8 +237,32 @@ def stencil_unknown_extent(self, arg, var_accesses=None):
+        var_name = DynStencils.dofmap_size_name(self._kern.root.symbol_table,
+                                                arg)
+        name = "{0}({1})".format(var_name,
+                                     self._cell_ref_name(var_accesses))
```
I suspect pylint will complain about the indentation of this line and 264.

andrewcoughtrie replied:
Fixed

arporter commented on src/psyclone/domain/lfric/kern_call_arg_list.py:
```diff
@@ -279,6 +303,27 @@ def stencil(self, arg, var_accesses=None):
+        from psyclone.dynamo0p3 import DynStencils
+        var_name = DynStencils.dofmap_name(self._kern.root.symbol_table, arg)
+        name = "{0}(:,:,:,{1})".format(var_name,
+                                     self._cell_ref_name(var_accesses))
```
Incorrect indent.

andrewcoughtrie replied:
Fixed.

arporter commented on src/psyclone/domain/lfric/kern_call_arg_list.py:
```diff
@@ -279,6 +303,27 @@ def stencil(self, arg, var_accesses=None):
+        # add in stencil dofmap
+        from psyclone.dynamo0p3 import DynStencils
+        var_name = DynStencils.dofmap_name(self._kern.root.symbol_table, arg)
+        name = "{0}(:,:,:,{1})".format(var_name,
```
Mainly to help my understanding: please could you explain why we're passing a 3D array for a 2D stencil (and similarly, I notice we're already passing a 2D array for a 1D stencil), even allowing for the fact that we need to index in to the stencil for the current cell. Either I'm missing something obvious or our documentation is a bit lacking.

andrewcoughtrie replied:
The array size passed for the stencil dofmaps is laid out like so: stencil_dofmap(dof in cell, cell in stencil, main cell) where main cell is obviously the cell the kernel is being called from the PSylayer, this is how the older stencils (e.g. cross, 1dx, 1dy, etc.) are passed. The problems with this (and the reason for the creation of the 2D cross stencil) is that when used on a planar mesh instead of a cubedsphere - as is the case with the limited area meshes - the size of the stencil at the boundaries changes, this is fine as long as you don't care which direction the stencil cell is in relation to the centre cell. If you do care about the direction (as is the case when calculating winds) you can't guarentee that a cell in the stencil is in a particular direction. To overcome this the 2D cross stencil bakes that information into the stencil dofmap, so: stencil_dofmap(dof in cell, cell in branch, branch in stencil, main cell) thus we have the extra dimension for the branch of the stencil which are always in the same order W,S,E,N. There are some pictures that might give you a better idea here: https://code.metoffice.gov.uk/trac/lfric/wiki/LFRicInfrastructure/StencilDOFMapDesign  

Hopefully this is helpful

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -1465,6 +1465,17 @@ def initialise(self, parent):
                                 ","+self.extent_value(arg)+")"))
                         parent.add(if_then)
+                elif stencil_type == "cross2d":
+                    # raise GenerationError(
```
(Oh, and in case you didn't know, adding the line `import pdb; pdb.set_trace()` will stop execution at the current line and open the Python debugger for you.)

andrewcoughtrie replied:
Thanks for the info, and done.

arporter commented on src/psyclone/tests/dynamo0p3_stubgen_test.py:
```diff
@@ -823,15 +823,15 @@ def test_stub_stencil_extent():
         "undf_w2, map_w2, ndf_w3, undf_w3, map_w3)")
     assert result1 in generated_code
-    result2 = "INTEGER(KIND=i_def), intent(in) :: field_2_stencil_size"
+    result2 = "INTEGER(KIND=i_def), intent(in) :: field_2_stencil_sizes"
```
It's a bit weird that this is now named "...sizes" plural but is still a single integer value. Is it possible to put this back?

andrewcoughtrie replied:
Unless I'm missing something this has to have the same name as the variables generated in the PSylayer as the name is pulled from the same methods. I could revert everything back to being ???_stencil_size. My original reason for changing the name was that it made more sense in the context of the variable in the PSylayer where they are now arrays of integers. I agree that in the kernels it may be more confusing and that is what the scientists see so.... Thoughts?

andrewcoughtrie replied:
I've changed everything back to ???_stencil_size.

arporter commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -4223,7 +4223,7 @@ def test_single_stencil_extent(dist_mem, tmpdir):
     assert output3 in result
     output4 = (
-        "      INTEGER(KIND=i_def) f2_stencil_size\n"
+        "      INTEGER(KIND=i_def), pointer :: f2_stencil_sizes(:) => null()\n"
```
I think it's OK to leave this name as `f2_stencil_size` - it's then more consistent with the naming of the other kernel arguments.

andrewcoughtrie replied:
I've reverted the name change to ???_stencil_size for everything, as you say it keeps things consistent and any possible naming oddness is kept to the PSylayer.

arporter commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -4374,6 +4374,53 @@ def test_stencil_region_unsupported(dist_mem):
+        "SUBROUTINE invoke_0_testkern_stencil_cross2d_type(f1, f2, f3, f4, "
+        "f2_extent)")
+    assert output1 in result
+    assert "USE stencil_2D_dofmap_mod, ONLY: stencil_2D_dofmap_type\n" in result
```
If you are able to merge those two if-blocks I mentioned earlier, this should become a single USE statement.

andrewcoughtrie replied:
Done

arporter commented on src/psyclone/tests/test_files/dynamo0p3/19.13_single_stencil.f90:
```diff
@@ -43,12 +43,12 @@ program single_stencil
+  integer(i_def)   :: f2_extent = 1, f3_stencil_sizes = 1
 
   call invoke(                                                    &
        testkern_stencil_type(f2_stencil_map, f2, f2_extent,       &
```
I think this was already wrong/incomplete but please could you add declarations for the kernel arguments. e.g. presumably `f2_stencil_map` is actually a field?

andrewcoughtrie replied:
Done, but that has to be the most confusing invoke statment ever.

arporter commented on src/psyclone/tests/test_files/dynamo0p3/19.25_single_stencil_cross2d.f90:
```diff
@@ -0,0 +1,50 @@
+! -----------------------------------------------------------------------------
+! BSD 3-Clause License
+!
+! Copyright (c) 2017-2020, Science and Technology Facilities Council
```
Just 2020 as this is a new file.

andrewcoughtrie replied:
Done.

arporter commented on src/psyclone/tests/test_files/dynamo0p3/testkern_stencil_cross2d_mod.f90:
```diff
@@ -0,0 +1,84 @@
+! -----------------------------------------------------------------------------
+! BSD 3-Clause License
+!
+! Copyright (c) 2017-2020, Science and Technology Facilities Council
```
Just 2020.

andrewcoughtrie replied:
Done.

arporter commented on src/psyclone/tests/test_files/dynamo0p3/testkern_stencil_cross2d_mod.f90:
```diff
@@ -0,0 +1,84 @@
+! OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+! -----------------------------------------------------------------------------
+! Author A. Coughtrie, Met Office
+
```
Please add a brief comment describing the purpose of this file.

andrewcoughtrie replied:
Done.

arporter reviewed (changes requested):
Good job. I was a bit confused about the code duplication in the class dealing with arguments for OpenACC but I think that's easily resolved (see inline).
I think the stencil size argument should still be named in the singular so that it is consistent with all of the other kernel arguments.
Trying to understand the changes here made me realise we probably don't have good enough documentation on the stencil side of things so a bit more of an explanation of the arrays that are passed would be good.
When running the test suite with compilation-checking enabled (--compile and --compileopencl flags to pytest) I get 56 failures. Probably because the chopped-down LFRic 'infrastructure' in the test suite (`tests/test_files/dynamo0p3/infrastructure`) needs updating to know about these changes to stencils.
I'll check for code conformance (pylint/pycodestyle) next time.

Just so you know, our working practice for a review is for you to respond to each comment (even if it's just 'done') and then it's up to the reviewer to decide whether or not that 'conversation' can be resolved at the next review.

andrewcoughtrie commented on src/psyclone/domain/lfric/kern_stub_arg_list.py:
```diff
@@ -32,7 +32,7 @@
 # -----------------------------------------------------------------------------
 # Authors R. W. Ford, A. R. Porter and S. Siso, STFC Daresbury Lab
-# Modified I. Kavcic, Met Office
+# Modified I. Kavcic and A. Coughtrie, Met Office
```
Done

andrewcoughtrie commented:
When dealing with the additions needed to the cut down LFRic infrastructure (which i'd completely forgotten to add) I came across an issue where an additional variable (max_branch_length) needed passing to the kernels when using the CROSS2D stencil. This is needed when defining size of the stencil_dofmap array due to the variable length of the branches at boundaries. The implementation of this is in commit fedc64a9a24b776236f0574e28337d72a303616a. 

I've updated the documentation to cover the missing info on stencils and the stencil_dofmaps passed into kernels.

arporter commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -215,6 +215,10 @@ def generate(self, var_accesses=None):
                             # metadata so must be passed
                             self.stencil_2d_unknown_extent \
                                 (arg, var_accesses=var_accesses)
+                            # Due to the nature of the stencil extent array
```
Having RTD, I think I now understand: the difference between a 'cross' and a 'cross2d' is that the branches of the latter may be of different lengths while in the former they are all of 'extent' length.
Therefore, the information passed from the Algorithm layer does not change but the kernel now needs to be supplied with the maximum branch length as well as the stencil size. So, section 11.1.8 can probably be left as is but a bit needs to be added on the Kernel interface. Naming wise, I appreciate that in LFRic, 'cross2d' has been chosen as the name so I guess we have to stick to that in PSyclone. I think the comments/docstrings referring to cross2d will have to be extended to make the meaning clear (e.g. "cross2d (irregular cross) stencil...")

rupertford replied:
It was a long time ago now but I think you are right @arporter. When first adding in support for stencils there was talk about needing stencil sizes that were not known until run-time but there was no example of this in LFRic at the time so I just implemented fixed sizes (with agreement from LFRic developers I believe) and added in metadata support ready for when they were needed.

andrewcoughtrie replied:
I've added some info to both 11.1.8 and the kernel interface to try and clarify why the stencil exists and how it is used/implemented.

arporter replied:
I've just created #963 to implement/unlock the ability to specify stencil extent in metadata.

arporter reviewed (changes requested):
Well, after getting myself thoroughly confused, I think I now understand. Apologies for my rambling comments. Please could you add a bit of the content from https://code.metoffice.gov.uk/trac/lfric/wiki/LFRicInfrastructure/StencilDOFMapDesign (with suitable attribution) to the section on stencils in the user guide. In fact, you need to add yourself to the author list once you've done that (doc/user_guide/conf.py).
Although I don't like the 'CROSS2D' name (all of our stencils are already 1D or 2D), I appreciate that the ship may have sailed on that one so I'm requesting that a bit of clarification be added in comments/docstrings where appropriate.

All tests and examples now pass, including with compilation.
Thanks for undoing your variable pluralisation.
New 'infrastructure' is nicely stripped down.

Almost there now. I'll check with pylint and pycodestyle next time.
Is there an LFRic branch that uses this new stencil functionality with which this PSyclone branch can be tested?

andrewcoughtrie commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -1424,11 +1424,23 @@ rules, along with PSyclone's naming conventions, are:
-         the number of cells in the stencil.
-      2) If the field entry stencil access is of type ``XORY1D`` then
-         add an integer direction argument with intent ``in``.
+         the number of cells in the stencil or in the case of the CROSS2D
```
Done

andrewcoughtrie commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -1424,11 +1424,23 @@ rules, along with PSyclone's naming conventions, are:
-         add an integer direction argument with intent ``in``.
+         the number of cells in the stencil or in the case of the CROSS2D
+         stencil the number of cells in each branch of the stencil.
+      2) If the stencil is of type CROSS2D then a integer of intent ``in``
```
Fixed

andrewcoughtrie commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -215,6 +215,10 @@ def generate(self, var_accesses=None):
                             # metadata so must be passed
```
Done

andrewcoughtrie commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -215,6 +215,10 @@ def generate(self, var_accesses=None):
                             # metadata so must be passed
                             self.stencil_2d_unknown_extent \
```
Done and for a couple of others.

andrewcoughtrie commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -447,6 +451,21 @@ def stencil_2d_unknown_extent(self, arg, var_accesses=None):
 
         '''
 
+    # @abc.abstractmethod
```
Uncommented

andrewcoughtrie commented on src/psyclone/domain/lfric/kern_call_arg_list.py:
```diff
@@ -166,12 +166,12 @@ def cma_operator(self, arg, var_accesses=None):
         components = ["matrix"]
         from psyclone.dynamo0p3 import DynCMAOperators
         if arg.function_space_to.orig_name != \
```
Done

andrewcoughtrie commented on src/psyclone/domain/lfric/kern_call_arg_list.py:
```diff
@@ -261,9 +260,29 @@ def stencil_2d_unknown_extent(self, arg, var_accesses=None):
         self.append(name, var_accesses, var_access_name=var_name)
 
+    def stencil_2d_max_extent(self, arg, var_accesses=None):
+        '''Add 2D stencil information to the argument list associated with the
```
Done

andrewcoughtrie commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -7124,9 +7140,11 @@ def test_dynkernelarguments_acc_args_4():
-        'f2_stencil_sizes', 'f2_stencil_dofmap', 'f3_proxy', 'f3_proxy%data',
-        'f4_proxy', 'f4_proxy%data', 'ndf_w1', 'undf_w1', 'map_w1', 'ndf_w2',
-        'undf_w2', 'map_w2', 'ndf_w3', 'undf_w3', 'map_w3']
+        'f2_stencil_size', 'f2_max_branch_length', 'f2_stencil_dofmap',
```
Yep that's correct :)

andrewcoughtrie commented on src/psyclone/dynamo0p3.py:
```diff
@@ -1466,16 +1489,18 @@ def initialise(self, parent):
+                        AssignGen(parent,
+                                  lhs=self.max_branch_length_name(symtab,
+                                                                  arg),
+                                  rhs=self.extent_value(arg) + " + 1_" +
```
The plus one is because the stencil branches contain the centre cell as well. As all stencils do.

andrewcoughtrie commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -447,6 +451,21 @@ def stencil_2d_unknown_extent(self, arg, var_accesses=None):
+    # @abc.abstractmethod
+    def stencil_2d_max_extent(self, arg, var_accesses=None):
+        '''Add 2D stencil information to the argument list associated with the
+        argument 'arg' if the maximum branch extent is unknown. If supplied it
```
This wouldn't be specified in the metadata as it's calculated from the stencil extent passed in from the Algorithm (or which could be in the metadata when the functionality is turned back on).

arporter replied:
So should this say "if the stencil extent (from which it is calculated) is passed from the Algorithm layer rather than being specified in kernel metadata"?

andrewcoughtrie commented:
I've tested the branch on a branch of LFRic (LFRic/branches/dev/andrewcoughtrie/r24298_stencil_port) on both GungHo with the test-suite to check the changes to the normal stencils work as well as testing with a miniapp (lam_dev) that implements a basic Algorithm/Kernel for the CROSS2D stencil and everything works as it should. 

Back to you @arporter

arporter commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -313,6 +313,17 @@ For example::
 
+If the stencil is of type ``CROSS2D`` then the arrays passed to the kernel are
+of different dimensions to those of other stencils. The ``CROSS2D`` stencil is
+designed for use when knowing where the stencil cells are in relation to the
```
"...for use when it is necessary for a kernel to know where the stencil cells are, relative to the current cell."

andrewcoughtrie replied:
Changed.

arporter reviewed (changes requested):
All tests pass, including with compilation.
Developer has tested branch with LFRic.
All examples OK (incl. compilation).
Updated UG builds fine.
Building the Reference Guide generates warnings but I don't think they're the result of this issue.
I've checked all changes with pylint and pycodestyle.

Just a few minor things to fix now.
Please could you also check that examples/lfric/eg5 is still correct and update if necessary.

andrewcoughtrie commented on src/psyclone/domain/lfric/kern_call_arg_list.py:
```diff
@@ -165,14 +165,14 @@ def cma_operator(self, arg, var_accesses=None):
-        if arg.function_space_to.orig_name != \
-           arg.function_space_from.orig_name:
+        if arg.function_space_to.orig_name != (arg.function_space_from.
+                orig_name):
```
Fixed the indenting

andrewcoughtrie commented on src/psyclone/domain/lfric/kern_call_acc_arg_list.py:
```diff
@@ -113,6 +113,57 @@ def stencil(self, arg, var_accesses=None):
+
+        '''
+        # The extent is not specified in the metadata so pass the value in
+        from psyclone.dynamo0p3 import DynStencils
```
I tried to do this but the tests all fall over with so many errors I can't work out what is going on. As soon as i put the import for DynStencils at the top of the file it seems to break eveything.

arporter replied:
That'll be down to circular dependencies. In which case, please just put a comment here to say that it has to be here to avoid circular dependencies and disable the pylint warning (search for pylint in the source to see how).

andrewcoughtrie commented on src/psyclone/domain/lfric/kern_call_arg_list.py:
```diff
@@ -237,7 +237,50 @@ def stencil_unknown_extent(self, arg, var_accesses=None):
+
+        '''
+        # The extent is not specified in the metadata so pass the value in
+        from psyclone.dynamo0p3 import DynStencils
```
Same problem as above.

andrewcoughtrie commented:
I've fixed the indent issue and had a look at the example, everything looks fine to me, the psylayer code is generated correctly and the kernel arguments are all good. Added the added the comments regarding circular dependencies and the pylint disable. Back to you @arporter

andrewcoughtrie commented:
Just noticed i missed a couple of your comments @arporter on the user guide will address those now.

andrewcoughtrie commented:
Right made those missed changes now so lets try this again :D @arporter

arporter commented:
All requested changes have been made.
All documentation builds OK.
All tests pass/xfail with compilation.
Examples run OK with compilation.
I've brought the branch up-to-date with master.
